File: cs/3_1_2_payment_strategy.cs
using System;
using System.Collections.Generic;

namespace Behavioral.PaymentStrategy
{
    class Card
    {
        public string Number;
        public string System;
        public DateTime Expires;
        public double Balance;
        public Card(string System, string Number, DateTime Expires, double Balance)
        {
            this.Number = Number;
            this.System = System;
            this.Expires = Expires;
            this.Balance = Balance;
        }
    }
    class MasterCard : Card
    {
        public MasterCard(string Number, DateTime Expires, double Balance) :
        base("MASTER", Number, Expires, Balance)
        { }
    }

    class Visa : Card
    {
        public Visa(string Number, DateTime Expires, double Balance) :
        base("VISA", Number, Expires, Balance)
        { }
    }

    class Bill
    {
        public double Sum;
        public Bill(double Sum)
        {
            this.Sum = Sum;
        }
    }
    interface IPayment
    {
        public bool Pay(double PaySum, Card CreditCard);
    }

    class PaymentProcessor
    {
        public Dictionary<string, IPayment> strategies;
        public PaymentProcessor()
        { }

        public bool Checkout(Bill bill, Card card)
        {
            if (!strategies.ContainsKey(card.System))
                return false;
            IPayment paymentStrategy = strategies.GetValueOrDefault(card.System);
            return paymentStrategy.Pay(bill.Sum, card);
        }
    }

    class VisaPayment : IPayment
    {
        public bool Pay(double PaySum, Card CreditCard)
        {
            if (CreditCard.System != "VISA")
            {
                Console.WriteLine("Wrong System");
                return false;
            }
            if (DateTime.Now > CreditCard.Expires)
            {
                Console.WriteLine("Expired");
                return false;
            }
            if (PaySum > CreditCard.Balance)
            {
                Console.WriteLine("Not enough money");
                return false;
            }
            CreditCard.Balance -= PaySum;
            Console.WriteLine("Payment by Visa");
            return true;
        }
    }

    class MasterCardPayment : IPayment
    {
        public bool Pay(double PaySum, Card CreditCard)
        {
            if (CreditCard.System != "MASTER")
            {
                Console.WriteLine("Wrong System");
                return false;
            }
            if (PaySum > CreditCard.Balance)
            {
                Console.WriteLine("Not enough money");
                return false;
            }
            if (DateTime.Now >= CreditCard.Expires)
            {
                Console.WriteLine("Expired");
                return false;
            }
            CreditCard.Balance -= PaySum;
            Console.WriteLine("Payment by MasterCard");
            return true;
        }
    }
}

File: python/patterns/_3_1_2_payment_strategy.py
from __future__ import annotations
from datetime import date
from decimal import Decimal
from typing import Protocol
from dataclasses import dataclass, field


@dataclass
class Card:
    number: str
    system: str
    expires: date
    balance: Decimal = Decimal("0")


@dataclass
class MasterCard (Card):
    system: str = field(init=False, default="MASTER")

@dataclass
class Visa (Card):
    system: str = field(init=False, default="VISA")


@dataclass
class Bill:
    sum: Decimal


class IPayment(Protocol):
    def pay(self, pay_sum: Decimal, credit_card: Card) -> bool: ...


class PaymentProcessor:
    def __init__(self) -> None:
        self.strategies: dict[str, IPayment] = {}
    

    def checkout(self, bill: Bill, card: Card) -> bool:
        if not card.system in self.strategies:
            return False
        return self.strategies[card.system].pay(bill.sum, card)


class VisaPayment:  # (IPayment)
    def pay(self, pay_sum: Decimal, credit_card: Card) -> bool:
        if credit_card.system != "VISA":
            print("Wrong System")
            return False

        if date.today() > credit_card.expires:
            print("Expired")
            return False

        if pay_sum > credit_card.balance:
            print("Not enough money")
            return False

        credit_card.balance -= pay_sum
        print("Payment by Visa")
        return True


class MasterCardPayment:  # (IPayment)
    def pay(self, pay_sum: Decimal, credit_card: Card) -> bool:
        if credit_card.system != "MASTER":
            print("Wrong System")
            return False

        if date.today() > credit_card.expires:
            print("Expired")
            return False

        if pay_sum > credit_card.balance:
            print("Not enough money")
            return False

        credit_card.balance -= pay_sum
        print("Payment by MasterCard")
        return True
